src/mds/Server.cc

```diff
@@ -8008,8 +7929,20 @@ void Server::handle_client_unlink(MDRequestRef& mdr)
   if (!dn)
     return;
 
-  if (is_reintegrate_pending(dn)) {
-    wait_for_pending_reintegrate(dn, mdr);
+  CDentry::linkage_t *dnl = dn->get_linkage(client, mdr);
+  ceph_assert(!dnl->is_null());
+  CInode *in = dnl->get_inode();
+
+  // wait the previous unlink finshes thoroughly.
+  CDentry* _dn = in->get_projected_parent_dn();
+  if (_dn->get_dir()->inode->is_stray()) {
+    mds->locker->drop_locks(mdr.get());
+
+    // after reintegrating retry it
+    in->add_waiter(CInode::WAIT_UNLINK, new C_MDS_RetryRequest(mdcache, mdr));
+
+    // trigger to flush the logs, after which will it trigger stray dentry reintegrate
+    mdlog->flush();
```

Comment:
Okay, hypothetical:

- MDS has nothing in cache
- client opens hard linked file, the primary dentry
- it unlinks that dentry, the remote dentry linking to it is not in cache so no stray migration is triggered
- optional: wait for journal flush so inode is no longer projected
- client tries to unlink the remote dentry (found via `rdlock_path_xlock_dentry`)

Will stray reintegration happen so that the unlink will eventually complete? (I ask because you believe that `rdlock_path_xlock_dentry` will not trigger stray reintegration.)

This type of situation for the code is not uncommon (it's one reason we have request deadlock in the MDS). I'm always leery of seeing a request wait like this without being certain some outstanding operation (which we can clearly point to) will wake it up.

Reply:
To be clear: what I'm hoping to see is some code path we can be certain will be hit when the `CInode` is no longer projected. How can we be sure stray reintegration will be triggered (if not already)?

Reply:
The `link` request is doing the same with here, please see https://github.com/ceph/ceph/blob/main/src/mds/Server.cc#L7385-L7396.

The `mdlog` flushing just to make sure that the previous `unlink` request's journal will be flushed and then the `_unlink_local() --> _unlink_local_finish() --> mdcache->notify_stray(straydn) --> ... --> StrayManager::_eval_stray() --> StrayManager::_eval_stray_remote() --> reintegrate_stray()` will be called.

And in `StrayManager::_eval_stray_remote()` it will pick a `remote` dentry arbitrarily:

```
640   /* If no remote_dn hinted, pick one arbitrarily */
641   if (remote_dn == NULL) {
642     if (!stray_in->remote_parents.empty()) {                                                                                                                                                                                       
643       for (const auto &dn : stray_in->remote_parents) {
644         if (dn->last == CEPH_NOSNAP && !dn->is_projected()) {
645           if (dn->is_auth()) { 
646             remote_dn = dn;    
647             if (remote_dn->dir->can_auth_pin()) 
648               break;
649           } else if (!remote_dn) {
650             remote_dn = dn;    
651           }
652         }
653       }
654     }
```

The reintegration will load the remote dentry when needed. And after the reintegration finishing it will wake up the waiters.

I know maybe the reintegration could be triggered by other cases, but that won't be our concern here. Isn't it ?